File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/pojo/AbstractRobot.java
package cz.muni.fi.umimecesky.pojo;

import android.view.ViewPropertyAnimator;
import android.widget.ImageView;

import java.util.concurrent.atomic.AtomicBoolean;

import cz.muni.fi.umimecesky.utils.CalculateDp;
import cz.muni.fi.umimecesky.utils.Constant;
import cz.muni.fi.umimecesky.utils.Global;

public abstract class AbstractRobot {

    private Runnable runnableBefore;
    private Runnable runnableAfter;

    public static AtomicBoolean noWinnerYet;

    private int hopsPerCorrect = 1;
    private volatile boolean lastWasCorrect = true;
    private int remainingHopsToWin;
    private ViewPropertyAnimator animator;

    private final float ROBOT_MOVE;
    private final int TOTAL_MOVES_TO_WIN;

    public abstract void applyCorrect();

    public abstract void applyWrong();

    public AbstractRobot(ImageView view, int hopsPerCorrect) {
        this.hopsPerCorrect = hopsPerCorrect;

        animator = view.animate();
        animator.setDuration(Constant.ANIMATION_DURATION);

        CalculateDp calculateDp = Global.getCalculateDp();
        ROBOT_MOVE = calculateDp.calculateRobotMovePx();

        TOTAL_MOVES_TO_WIN = calculateDp.getWinMovesCount();
        remainingHopsToWin = TOTAL_MOVES_TO_WIN;
    }

    protected final void moveForward() {
        lastWasCorrect = true;
        if (isOneStepToWin()) {
            decideAnimationForward();
        } else if (noWinnerYet.get()) {
            decideAnimationForward();
        }
    }

    private void decideAnimationForward() {
        if (isOneStepToWin()) {
            animateForwardWithAction(runnableBefore, runnableAfter);
        } else {
            animateForward();
        }
        remainingHopsToWin -= limitedHopsPerCorrect();
    }

    protected final void moveBackward() {
        if (lastWasCorrect) {
            lastWasCorrect = false;

            if (isNotAtStart()) {
                animateBackward();
            }
        }
    }

    private void animateBackward() {
        animator.translationXBy(-ROBOT_MOVE);
        remainingHopsToWin++;
    }

    private void animateForward() {
        animateForwardWithAction(null, null);
    }

    protected void animateForwardWithAction(Runnable runnableBefore, Runnable runnableAfter) {
        animator.translationXBy(ROBOT_MOVE * limitedHopsPerCorrect())
                .withStartAction(runnableBefore)
                .withEndAction(runnableAfter);
    }

    private int limitedHopsPerCorrect() {
        return limitHopsWhenWon(hopsPerCorrect);
    }

    private int limitHopsWhenWon(int hopsBefore) {
        if (remainingHopsToWin >= hopsBefore) {
            return hopsBefore;
        } else {
            return remainingHopsToWin;
        }

    }

    protected void setLastWasCorrect(boolean lastWasCorrect) {
        this.lastWasCorrect = lastWasCorrect;
    }

    protected boolean getLastWasCorrect() {
        return lastWasCorrect;
    }

    private boolean isOneStepToWin() {
        return 0 < remainingHopsToWin && remainingHopsToWin <= hopsPerCorrect;
    }

    private boolean isNotAtStart() {
        return remainingHopsToWin != TOTAL_MOVES_TO_WIN;
    }

    public void setRunnableAfter(Runnable runnableAfter) {
        this.runnableAfter = runnableAfter;
    }

    public void setRunnableBefore(Runnable runnableBefore) {
        this.runnableBefore = runnableBefore;
    }
}

File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/pojo/AbstractRobot.kt
package cz.muni.fi.umimecesky.pojo

import android.view.ViewPropertyAnimator
import android.widget.ImageView
import cz.muni.fi.umimecesky.utils.Constant
import cz.muni.fi.umimecesky.utils.Global
import java.util.concurrent.atomic.AtomicBoolean

abstract class AbstractRobot(view: ImageView, val hopsPerCorrect: Int) {

    private var runnableBefore: Runnable? = null
    private var runnableAfter: Runnable? = null

    @Volatile protected var lastWasCorrect = true
    private var remainingHopsToWin: Int = 0
    private val animator: ViewPropertyAnimator = view.animate()

    private val robotMove: Float
    private val totalMovesToWin: Int

    abstract fun applyCorrect()

    abstract fun applyWrong()

    init {
        animator.duration = Constant.ANIMATION_DURATION.toLong()

        val calculateDp = Global.calculateDp
        robotMove = calculateDp.calculateRobotMovePx()

        totalMovesToWin = calculateDp.winMovesCount
        remainingHopsToWin = totalMovesToWin
    }

    protected fun moveForward() {
        lastWasCorrect = true
        if (isOneStepToWin) {
            decideAnimationForward()
        } else if (noWinnerYet.get()) {
            decideAnimationForward()
        }
    }

    private fun decideAnimationForward() {
        if (isOneStepToWin) {
            animateForwardWithAction(runnableBefore, runnableAfter)
        } else {
            animateForward()
        }
        remainingHopsToWin -= limitedHopsPerCorrect()
    }

    protected fun moveBackward() {
        if (lastWasCorrect) {
            lastWasCorrect = false

            if (isNotAtStart) {
                animateBackward()
            }
        }
    }

    private fun animateBackward() {
        animator.translationXBy(-robotMove)
        remainingHopsToWin++
    }

    private fun animateForward() {
        animateForwardWithAction(null, null)
    }

    private fun animateForwardWithAction(runnableBefore: Runnable?, runnableAfter: Runnable?) {
        animator.translationXBy(robotMove * limitedHopsPerCorrect())
                .withStartAction(runnableBefore)
                .withEndAction(runnableAfter)
    }

    private fun limitedHopsPerCorrect(): Int = limitHopsWhenWon(hopsPerCorrect)

    private fun limitHopsWhenWon(hopsBefore: Int): Int =
            if (remainingHopsToWin >= hopsBefore) hopsBefore else remainingHopsToWin

    private val isOneStepToWin: Boolean
        get() = remainingHopsToWin in 1..hopsPerCorrect

    private val isNotAtStart: Boolean
        get() = remainingHopsToWin != totalMovesToWin

    fun setRunnableAfter(runnableAfter: Runnable) {
        this.runnableAfter = runnableAfter
    }

    fun setRunnableBefore(runnableBefore: Runnable) {
        this.runnableBefore = runnableBefore
    }

    companion object {
        var noWinnerYet = AtomicBoolean(true)
    }
}
